programs/whirlpool/src/util/sparse_swap.rs

```diff
@@ -0,0 +1,1746 @@
+
+enum TickArrayAccount<'info> {
+    Initialized {
+        tick_array_whirlpool: Pubkey,
+        start_tick_index: i32,
+        account_info: AccountInfo<'info>,
+    },
+    Uninitialized {
+        pubkey: Pubkey,
+        account_info: AccountInfo<'info>,
+        start_tick_index: Option<i32>,
+    },
+}
+
+pub struct SparseSwapTickSequenceBuilder<'info> {
+    tick_array_accounts: Vec<TickArrayAccount<'info>>,
+}
+
+impl<'info> SparseSwapTickSequenceBuilder<'info> {
+    pub fn try_from(
+        whirlpool: &Account<'info, Whirlpool>,
+        a_to_b: bool,
+        static_tick_array_account_infos: Vec<AccountInfo<'info>>,
+        supplemental_tick_array_account_infos: Option<Vec<AccountInfo<'info>>>,
+    ) -> Result<Self> {
+        let mut tick_array_account_infos = static_tick_array_account_infos;
+        if let Some(supplemental_tick_array_account_infos) = supplemental_tick_array_account_infos {
+            tick_array_account_infos.extend(supplemental_tick_array_account_infos);
+        }
+
+        // dedup by key
+        tick_array_account_infos.sort_by_key(|a| a.key());
+        tick_array_account_infos.dedup_by_key(|a| a.key());
+
+        let mut initialized = vec![];
+        let mut uninitialized = vec![];
+        for account_info in tick_array_account_infos.into_iter() {
+            let state = peek_tick_array(account_info)?;
+
+            match &state {
+                TickArrayAccount::Initialized {
+                    tick_array_whirlpool,
+                    start_tick_index,
+                    ..
+                } => {
+                    // has_one constraint equivalent check
+                    if *tick_array_whirlpool != whirlpool.key() {
+                        return Err(ErrorCode::DifferentWhirlpoolTickArrayAccount.into());
+                    }
+
+                    // TickArray accounts in initialized have been verified as:
+                    //   - Owned by this program
+                    //   - Initialized as TickArray account
+                    //   - Writable account
+                    //   - TickArray account for this whirlpool
+                    // So we can safely use these accounts.
+                    initialized.push((*start_tick_index, state));
+                }
+                TickArrayAccount::Uninitialized {
+                    pubkey: account_address,
+                    ..
+                } => {
+                    // TickArray accounts in uninitialized have been verified as:
+                    //   - Owned by System program
+                    //   - Data size is zero
+                    //   - Writable account
+                    // But we are not sure if these accounts are valid TickArray PDA for this whirlpool,
+                    // so we need to check it later.
+                    uninitialized.push((*account_address, state));
+                }
+            }
+        }
+
+        let start_tick_indexes = get_start_tick_indexes(whirlpool, a_to_b);
```

Comment:
Added comment.👍

PDA calculation is a high cost process that costs 3000 CU/PDA ~, so if it can be matched against start_tick_index, it is preferred.